File: app/src/main/java/com/jzoft/ygohelper/biz/ProxyCardHolder.java
package com.jzoft.ygohelper.biz;

import com.jzoft.ygohelper.utils.HttpCaller;

import java.util.LinkedList;
import java.util.List;

/**
 * Created by jjimenez on 11/10/16.
 */
public class ProxyCardHolder {
    private final ProxyCardLocator localizator;
    private List<ProxyCard> proxyCards;

    public ProxyCardHolder(ProxyCardLocator localizator) {
        this.localizator = localizator;
        proxyCards = new LinkedList<ProxyCard>();
    }

    public void copy(int index) {
        try {
            proxyCards.add(index + 1, proxyCards.get(index));
        } catch (IndexOutOfBoundsException e) {
            throw new IllegalArgumentException();
        }
    }

    public void add(String location) throws HttpCaller.NotFound {
        ProxyCard proxyCard = localizator.locate(location);
        if (proxyCard == null)
            throw new IllegalArgumentException();
        proxyCards.add(proxyCard);
    }

    public void remove(int index) {
        proxyCards.remove(index);
    }

    public List<ProxyCard> getProxyCards() {
        return proxyCards;
    }
}

File: app/src/main/java/com/jzoft/ygohelper/biz/ProxyCardHolder.kt
package com.jzoft.ygohelper.biz

import com.jzoft.ygohelper.utils.Caller

import java.util.LinkedList

/**
 * Created by jjimenez on 11/10/16.
 */
class ProxyCardHolder(private val localizator: ProxyCardLocator) {
    val proxyCards: MutableList<ProxyCard>

    init {
        proxyCards = LinkedList()
    }

    fun copy(index: Int) {
        try {
            proxyCards.add(index + 1, proxyCards[index])
        } catch (e: IndexOutOfBoundsException) {
            throw IllegalArgumentException()
        }

    }

    @Throws(Caller.NotFound::class)
    fun add(location: String) {
        localizator.locate(location).subscribe { proxyCards.add(it) }
    }

    fun remove(index: Int) {
        proxyCards.removeAt(index)
    }
}
